File: zig/day20.zig
const std = @import("std");
const runDay = @import("adventofcode2023.zig").runDay;

const Allocator = std.mem.Allocator;

const TEST_CASE =
    \\broadcaster -> a, b, c
    \\%a -> b
    \\%b -> c
    \\%c -> inv
    \\&inv -> a
;

const TEST_CASE2 =
    \\broadcaster -> a
    \\%a -> inv, con
    \\&inv -> b
    \\%b -> con
    \\&con -> output
;

const Pulse = enum { high, low };
const ModuleKind = enum { Broadcast, FlipFlop, Conjunction, Untyped };
const ModuleBroadcast = struct {
    outputs: std.ArrayList([]const u8),

    fn receive(self: *ModuleBroadcast, pulse: Pulse, source: []const u8, modules: *ModuleMap) void {
        _ = source;

        for (self.*.outputs.items) |output| {
            modules.send(output, pulse, "broadcaster");
        }
    }

    fn print(self: *const ModuleBroadcast) !void {
        _ = self;
    }
};

const ModuleUntyped = struct {
    name: []const u8,

    fn receive(self: *ModuleUntyped, pulse: Pulse, source: []const u8, modules: *ModuleMap) void {
        _ = modules;
        std.debug.print("{s}: Received {} from {s}\n", .{ self.*.name, pulse, source });
    }

    fn print(self: *const ModuleUntyped) !void {
        _ = self;
    }
};

const ModuleFlipFlop = struct {
    name: []const u8,
    state: bool,
    outputs: std.ArrayList([]const u8),

    fn new(name: []const u8, outputs: std.ArrayList([]const u8)) ModuleFlipFlop {
        return ModuleFlipFlop{ .name = name, .state = false, .outputs = outputs };
    }

    fn receive(self: *ModuleFlipFlop, pulse: Pulse, source: []const u8, modules: *ModuleMap) void {
        _ = source;

        if (pulse == Pulse.high) {
            return;
        }

        self.state = !self.state;
        const next_pulse = switch (self.state) {
            true => Pulse.high,
            false => Pulse.low,
        };

        for (self.*.outputs.items) |output| {
            modules.send(output, next_pulse, self.*.name);
        }
    }

    fn print(self: *const ModuleFlipFlop) !void {
        std.debug.print("state={}", .{self.*.state});
    }

    fn deinit(self: Module) void {
        self.outputs.deinit();
    }
};

const ModuleConjunction = struct {
    name: []const u8,
    inputs: std.StringHashMap(Pulse),
    outputs: std.ArrayList([]const u8),

    fn new(allocator: Allocator, name: []const u8, outputs: std.ArrayList([]const u8)) !ModuleConjunction {
        const inputs = std.StringHashMap(Pulse).init(allocator);

        return ModuleConjunction{ .name = name, .inputs = inputs, .outputs = outputs };
    }

    fn receive(self: *ModuleConjunction, pulse: Pulse, source: []const u8, modules: *ModuleMap) void {
        self.*.inputs.put(source, pulse) catch unreachable;

        var all_high = true;
        var it = self.*.inputs.valueIterator();
        while (it.next()) |val| {
            all_high = all_high and (val.* == Pulse.high);
        }

        const next_pulse = switch (all_high) {
            true => Pulse.low,
            false => Pulse.high,
        };

        for (self.*.outputs.items) |output| {
            modules.send(output, next_pulse, self.*.name);
        }
    }

    fn print(self: *const ModuleConjunction) !void {
        std.debug.print("inputs=(", .{});
        var it = self.*.inputs.iterator();
        while (it.next()) |kv| {
            std.debug.print("{s}={}, ", .{ kv.key_ptr.*, kv.value_ptr.* });
        }
        std.debug.print(")", .{});
    }

    fn deinit(self: Module) void {
        self.inputs.deinit();
        self.outputs.deinit();
    }
};

const Module = union(ModuleKind) {
    Broadcast: ModuleBroadcast,
    FlipFlop: ModuleFlipFlop,
    Conjunction: ModuleConjunction,
    Untyped: ModuleUntyped,

    fn new_untyped(m_name: []const u8) Module {
        return Module{ .Untyped = ModuleUntyped{ .name = m_name } };
    }

    fn parse(allocator: Allocator, input: []const u8) !Module {
        var it = std.mem.split(u8, input, " -> ");
        var m_name = it.next().?;

        var m_outputs = std.ArrayList([]const u8).init(allocator);
        var it_outputs = std.mem.split(u8, it.next().?, ", ");
        while (it_outputs.next()) |o| {
            try m_outputs.append(o);
        }

        switch (m_name[0]) {
            '%' => return Module{ .FlipFlop = ModuleFlipFlop.new(m_name[1..], m_outputs) },
            '&' => return Module{ .Conjunction = try ModuleConjunction.new(allocator, m_name[1..], m_outputs) },
            else => if (std.mem.eql(u8, m_name, "broadcaster")) {
                return Module{ .Broadcast = ModuleBroadcast{ .outputs = m_outputs } };
            } else {
                return Module{ .Untyped = ModuleUntyped{ .name = m_name } };
            },
        }
    }

    fn name(self: *const Module) []const u8 {
        return switch (self.*) {
            .Broadcast => |_| "broadcaster",
            .Untyped => |m| m.name,
            .FlipFlop => |m| m.name,
            .Conjunction => |m| m.name,
        };
    }

    fn outputs(self: *const Module) [][]const u8 {
        return switch (self.*) {
            .Broadcast => |m| m.outputs.items,
            .Untyped => |_| &[_][]const u8{},
            .FlipFlop => |m| m.outputs.items,
            .Conjunction => |m| m.outputs.items,
        };
    }

    fn print(self: *const Module) void {
        std.debug.print("{}({s}) -> ", .{ @as(ModuleKind, self.*), self.*.name() });
        for (self.*.outputs()) |item| {
            std.debug.print("{s}, ", .{item});
        }
        try switch (self.*) {
            .Broadcast => |m| m.print(),
            .Untyped => |m| m.print(),
            .FlipFlop => |m| m.print(),
            .Conjunction => |m| m.print(),
        };
        std.debug.print("\n", .{});
    }

    fn receive(self: *Module, pulse: Pulse, source: []const u8, modules: *ModuleMap) void {
        switch (self.*) {
            .FlipFlop => |*m| m.receive(pulse, source, modules),
            .Conjunction => |*m| m.receive(pulse, source, modules),
            .Broadcast => |*m| m.receive(pulse, source, modules),
            .Untyped => |*m| m.receive(pulse, source, modules),
        }
        modules.flush();
    }
};

const SentPulse = struct {
    output: []const u8,
    pulse: Pulse,
    source: []const u8,
};

const ModuleMap = struct {
    allocator: Allocator,
    modules: std.StringHashMap(Module),
    queue: std.ArrayList(SentPulse),
    count_low: usize = 0,
    count_high: usize = 0,

    fn new(allocator: Allocator) ModuleMap {
        return ModuleMap{ .allocator = allocator, .modules = std.StringHashMap(Module).init(allocator), .queue = std.ArrayList(SentPulse).init(allocator) };
    }

    fn parse(allocator: Allocator, input: []const u8) !ModuleMap {
        var self = ModuleMap.new(allocator);
        var it = std.mem.split(u8, input, "\n");
        while (it.next()) |row| {
            const module = try Module.parse(allocator, row);
            try self.modules.put(module.name(), module);
        }

        return self;
    }

    fn send(self: *ModuleMap, output: []const u8, pulse: Pulse, source: []const u8) void {
        std.debug.print("{s} -{s}-> {s}\n", .{ source, switch (pulse) {
            .high => "high",
            .low => "low",
        }, output });

        if (pulse == Pulse.high) {
            self.*.count_high += 1;
        } else {
            self.*.count_low += 1;
        }

        self.get(output).?.receive(pulse, source, self);
        self.queue.append(SentPulse{ .output = output, .pulse = pulse, .source = source }) catch unreachable;
    }

    fn flush(self: *ModuleMap) void {
        while (self.*.queue.items.len > 0) {
            var q = self.*.queue;
            defer q.deinit();

            self.*.queue = std.ArrayList(SentPulse).init(self.allocator);

            for (q.items) |sent_pulse| {
                self.get(sent_pulse.output).?.receive(sent_pulse.pulse, sent_pulse.source, self);
            }
        }
    }

    fn get(self: *ModuleMap, name: []const u8) ?*Module {
        if (self.*.modules.getPtr(name)) |module| {
            return module;
        } else {
            self.*.modules.put(name, Module.new_untyped(name)) catch unreachable;
            return self.*.modules.getPtr(name).?;
        }
    }

    fn print(self: *const ModuleMap) void {
        var it = self.*.modules.valueIterator();
        while (it.next()) |module| {
            module.print();
        }
        std.debug.print("{} low and {} high: {}\n", .{ self.*.count_low, self.*.count_high, self.result_value() });
    }

    fn result_value(self: *const ModuleMap) usize {
        return self.*.count_low * self.*.count_high;
    }

    fn deinit(self: ModuleMap) void {
        var it = self.modules.valuesIterator();

        while (it.next()) |module| {
            module.deinit();
        }
        self.modules.deinit();
        self.queue.deinit();
    }
};

test "parse p1" {
    std.testing.refAllDecls(@This());

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    {
        var map = try ModuleMap.parse(allocator, TEST_CASE);
        for (0..1000) |_| {
            map.send("broadcaster", Pulse.low, "button");
            map.flush();
        }
        map.print();
    }
    {
        var map = try ModuleMap.parse(allocator, TEST_CASE2);
        for (0..4) |i| {
            std.debug.print("\n\n --- {}\n\n", .{i});
            map.send("broadcaster", Pulse.low, "button");
            map.flush();
        }
        map.print();
    }
}

const Day = struct {
    pub fn run1(allocator: std.mem.Allocator, input: []const u8) u64 {
        _ = input;
        _ = allocator;

        return 0;
    }

    pub fn run2(allocator: std.mem.Allocator, input: []const u8) u64 {
        _ = input;
        _ = allocator;

        return 0;
    }
};

pub fn main() !void {
    try runDay(Day, 0);
}

File: rust/src/bin/day20.rs
use adventofcode2023::AocSolution;

use std::{
    cell::RefCell,
    collections::{HashMap, VecDeque},
};

use itertools::Itertools;

pub struct Solution;

#[derive(Clone, Copy, Debug)]
enum Pulse {
    High,
    Low,
}

#[derive(Clone, Copy, Debug)]
struct PulseSignal<'a> {
    src: &'a str,
    dst: &'a str,
    pulse: Pulse,
}

#[derive(Debug)]
enum Module<'a> {
    Broadcaster {
        output: Vec<&'a str>,
    },
    FlipFlop {
        name: &'a str,
        output: Vec<&'a str>,
        state: bool,
    },
    Conjunction {
        name: &'a str,
        output: Vec<&'a str>,
        inputs: HashMap<String, Pulse>,
    },
    Untyped {
        name: &'a str,
    },
}

impl<'a> Module<'a> {
    fn parse(input: &'a str) -> (&'a str, Self) {
        let (name, output_names) = input.split(" -> ").next_tuple().unwrap();

        let output = output_names.split(", ").collect::<Vec<_>>();

        match name {
            "broadcaster" => ("broadcaster", Module::Broadcaster { output }),
            name if name.starts_with('%') => (
                &name[1..],
                Module::FlipFlop {
                    name: &name[1..],
                    output,
                    state: false,
                },
            ),
            name if name.starts_with('&') => (
                &name[1..],
                Module::Conjunction {
                    name: &name[1..],
                    inputs: HashMap::new(),
                    output,
                },
            ),
            _ => unreachable!(),
        }
    }

    fn output_names(&self) -> &[&'a str] {
        match self {
            Module::Broadcaster { output } => output,
            Module::FlipFlop { output, .. } => output,
            Module::Conjunction { output, .. } => output,
            Module::Untyped { .. } => &[],
        }
    }

    fn receive(&mut self, source: &'a str, pulse: Pulse) -> Option<Vec<PulseSignal<'a>>> {
        match self {
            Module::Broadcaster { output } => Some(
                output
                    .iter()
                    .map(|dst| PulseSignal {
                        src: "broadcaster",
                        dst,
                        pulse,
                    })
                    .collect(),
            ),
            Module::FlipFlop {
                name,
                output,
                state,
            } => {
                if let Pulse::High = pulse {
                    None
                } else {
                    *state = !*state;
                    let pulse = if *state { Pulse::High } else { Pulse::Low };
                    Some(
                        output
                            .iter()
                            .map(|dst| PulseSignal {
                                src: name,
                                dst,
                                pulse,
                            })
                            .collect(),
                    )
                }
            }
            Module::Conjunction {
                name,
                output,
                inputs,
            } => {
                inputs.insert(source.to_string(), pulse);
                let pulse = if inputs.values().all(|pulse| matches!(pulse, Pulse::High)) {
                    Pulse::Low
                } else {
                    Pulse::High
                };

                Some(
                    output
                        .iter()
                        .map(|dst| PulseSignal {
                            src: name,
                            dst,
                            pulse,
                        })
                        .collect(),
                )
            }
            Module::Untyped { name: _ } => None,
        }
    }
}

#[derive(Debug)]
struct ModuleMap<'a> {
    modules: RefCell<HashMap<String, Module<'a>>>,
    count_high: u64,
    count_low: u64,
    count_rx: u64,
}

impl<'a> ModuleMap<'a> {
    fn parse(input: &'a str) -> Self {
        let mut module_map = input
            .trim()
            .lines()
            .map(Module::parse)
            .map(|(k, v)| (k.to_string(), v))
            .collect::<HashMap<_, _>>();

        let input_names = module_map
            .iter()
            .flat_map(|(name, module)| {
                module.output_names().iter().filter_map(|&dst_name| {
                    if matches!(module_map.get(dst_name), Some(Module::Conjunction { .. })) {
                        Some((name.to_string(), dst_name.to_string()))
                    } else {
                        None
                    }
                })
            })
            .collect_vec();

        for (input_name, conj_name) in input_names {
            if let Some(Module::Conjunction { inputs, .. }) = module_map.get_mut(conj_name.as_str())
            {
                inputs.insert(input_name, Pulse::Low);
            } else {
                panic!();
            }
        }

        Self {
            modules: RefCell::new(module_map),
            count_low: 0,
            count_high: 0,
            count_rx: 0,
        }
    }

    fn push_button(&mut self) {
        self.count_rx = 0;

        let mut q = VecDeque::new();
        q.push_front(PulseSignal {
            src: "button",
            dst: "broadcaster",
            pulse: Pulse::Low,
        });

        while !q.is_empty() {
            let mut map = self.modules.borrow_mut();
            let signal = q.pop_front().unwrap();
            match signal.pulse {
                Pulse::High => self.count_high += 1,
                Pulse::Low => self.count_low += 1,
            }

            if matches!(signal.pulse, Pulse::Low) && signal.dst == "rx" {
                self.count_rx += 1;
            }

            if !map.contains_key(signal.dst) {
                map.insert(signal.dst.to_string(), Module::Untyped { name: signal.dst });
            }

            if let Some(signals) = map
                .get_mut(signal.dst)
                .and_then(|m| m.receive(signal.src, signal.pulse))
            {
                signals.into_iter().for_each(|s| q.push_back(s));
            }
        }
    }

    fn counts(&self) -> u64 {
        self.count_high * self.count_low
    }
}

impl AocSolution for Solution {
    const DAY: u8 = 20;

    fn new() -> Self {
        Self
    }

    fn part1(&self, input: &str) -> u64 {
        let mut mm = ModuleMap::parse(input);
        for _ in 0..1000 {
            mm.push_button();
        }
        mm.counts()
    }

    fn part2(&self, input: &str) -> u64 {
        let mut mm = ModuleMap::parse(input);
        (0u64..)
            .find(|i| {
                mm.push_button();
                if i % 10000 == 0 {
                    println!("{i} {}", mm.count_rx);
                }
                mm.count_rx == 1
            })
            .unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_CASE: &str = textwrap_macros::dedent!(
        r"
    broadcaster -> a, b, c
    %a -> b
    %b -> c
    %c -> inv
    &inv -> a
    "
    );

    const TEST_CASE2: &str = textwrap_macros::dedent!(
        r"
    broadcaster -> a
    %a -> inv, con
    &inv -> b
    %b -> con
    &con -> output
    "
    );

    #[test]
    fn test_parse() {
        let mut mm = ModuleMap::parse(TEST_CASE);
        for _ in 0..1000 {
            mm.push_button();
        }
        println!("{}x{} -> {}", mm.count_low, mm.count_high, mm.counts());

        let mut mm = ModuleMap::parse(TEST_CASE2);
        for _ in 0..1000 {
            mm.push_button();
        }
        println!("{}x{} -> {}", mm.count_low, mm.count_high, mm.counts());
    }
}

adventofcode2023::run!(Solution);
